feat: add support for optional field 'installedFrom'

coyoteecd opened the pull request:
This PR adds a "downloadLink" field that displays the "resolved" URL for each package, as extracted from package-lock.json.
Some legal departments require the "source" the package was downloaded from.

For example, `https://registry.npmjs.org/buffer/-/buffer-4.9.2.tgz`

BePo65 commented:
As we all know that we should not discuss with legal it is probably a good idea to add this field :-)
I think this is an excellent pr, it even includes the updated tests that are hard work to do.

With my heavy focus on CleanCode I would like to discuss about some of the chosen names for variables and functions.
My ideas are:
1. in 'config.js' we could
   rename 'DownloadLink' to 'sourceLink' and rename existing field 'link' to 'repositoryLink' (to differentiate the 2 'links' - breaking change!)
   or what I would prefer
   rename 'DownloadLink' to 'source' (and keep 'link'; no breaking change)

   In any case; the name of the new field must be camelCase to be consistent with the other fields.

2. in 'getInstalledVersions.js'
   rename the variable 'installedVersions' to 'installedPackagesData' (as this is what we extract from the file)
   rename the variable 'installedVersion' to 'installedPackageData'

   rename the file 'getInstalledVersions' to 'getInstalledPackagesData', as the function returns more than just the 'installed versions'.  This renaming should be done for the file and for the variable in index.js.

3. in 'getPackageReportData.js'
   Problem: if we use the names 'installedVersion' and 'installedVersionDownloadLink', one time we use 'version' as a synonym for 'version number' and one time as a synonym for 'package'; this is not consistent. To make the names more consistent, we could:
   rename 'installedVersion' to 'installedPackageVersion' and
   rename 'installedVersionDownloadLink' to 'installedPackageSource' (or what we did to 'DownloadLink' in 1.)
   or what I like more
   keep 'installedVersion' and
   rename 'installedVersionDownloadLink' to 'source' (or what we did to 'DownloadLink' in 1.)

   and in the returned object:
   rename 'downloadLink' to 'source' (or what we did to 'DownloadLink' in 1.)

4. consider the change from 1. to all the tests and test fixtures (sorry for that :-)

What do you think about all that?

BTW: I had to update the packages as a result of the last dependabot report. So it would be nice, if you could rebase this pr to the latest commit of master.

coyoteecd commented:
@BePo65 Could you use the inline review comments feature in GitHub (switch to File Changes tab and add them to the related changes). This leaves room for discussing them individually. For now, I will reply below

> As we all know that we should not discuss with legal it is probably a good idea to add this field :-) I think this is an excellent pr, it even includes the updated tests that are hard work to do.
> 
> With my heavy focus on CleanCode I would like to discuss about some of the chosen names for variables and functions. My ideas are:
> 
>     1. in 'config.js' we could
>        rename 'DownloadLink' to 'sourceLink' and rename existing field 'link' to 'repositoryLink' (to differentiate the 2 'links' - breaking change!)
>        or what I would prefer
>        rename 'DownloadLink' to 'source' (and keep 'link'; no breaking change)
>        In any case; the name of the new field must be camelCase to be consistent with the other fields.

First, `downloadLink` is already camelCase (this is where inline reviews help).
Second, I don't mind the change, but imo `source` is a bit unclear, is it a link, and if so, what's different from existing `link`.
How about `installedFrom`? Lines up well with installedVersion and no breaking changes...

> 
>     2. in 'getInstalledVersions.js'
>        rename the variable 'installedVersions' to 'installedPackagesData' (as this is what we extract from the file)
>        rename the variable 'installedVersion' to 'installedPackageData'
>        rename the file 'getInstalledVersions' to 'getInstalledPackagesData', as the function returns more than just the 'installed versions'.  This renaming should be done for the file and for the variable in index.js.

Yeah I wanted to avoid submitting way too many changes since some maintainers are not open to merging those.
If you're okay with accepting bigger changes I will follow with other features that we need in our project.

I will do the renames.

> 
>     3. in 'getPackageReportData.js'
>        Problem: if we use the names 'installedVersion' and 'installedVersionDownloadLink', one time we use 'version' as a synonym for 'version number' and one time as a synonym for 'package'; this is not consistent. To make the names more consistent, we could:
>        rename 'installedVersion' to 'installedPackageVersion' and
>        rename 'installedVersionDownloadLink' to 'installedPackageSource' (or what we did to 'DownloadLink' in 1.)
>        or what I like more
>        keep 'installedVersion' and
>        rename 'installedVersionDownloadLink' to 'source' (or what we did to 'DownloadLink' in 1.)
>        and in the returned object:
>        rename 'downloadLink' to 'source' (or what we did to 'DownloadLink' in 1.)

Will do

> 
>     4. consider the change from 1. to all the tests and test fixtures (sorry for that :-)

Feedback: the tests were a pain to modify (not fun to figure out why a large single-line html string does not equal another one)
I would suggest using less data, or formatting it in a friendlier way.

> 
> 
> What do you think about all that?
> 
> BTW: I had to update the packages as a result of the last dependabot report. So it would be nice, if you could rebase this pr to the latest commit of master.

coyoteecd commented:
@BePo65 I force-pushed a commit with the renames (see above).

BePo65 commented on README.md:
```diff
@@ -128,6 +128,7 @@ Fields with data of the installed packages:
 | name | name | name of the package |
 | licenseType | license type | type of the license of the package (e.g. MIT) |
 | link | link | link to the repository of the package |
+| installedFrom | installed from | the download source for the installed package |
```
yes, this is a much better name

BePo65 commented on lib/getInstalledPackagesData.js:
```diff
@@ -0,0 +1,37 @@
+module.exports = getInstalledPackagesData
+
+/**
+ * Get the installed versions from package-lock.json.
```
Should also mention the new fields extracted; something like:
Get several fields from package-lock.json like 'version' and 'resolved'.

BePo65 commented on lib/config.js:
```diff
@@ -63,6 +63,7 @@ module.exports = rc('license-report', {
 		'material',
 		'licenseType',
 		'link',
+		'installedFrom',
```
After reflecting for some more time about the 'fields' property I think that adding the new field to this list is a breaking change (as it changes the default output).

As not everybody might always need this field, we probably **should leave the new field out of 'fields'** array - or what is your opinion.

Doing this the user of license-report must add an argument to the command line to use 'installedFrom' (described as 'select fields for output' in the readme or by simply adding at least 2 arguments '--fields=<fieldname>' when calling 'node license-report').

And we could keep the e2e tests as they are, just adding one test for the complete field list (with json output to make generate the expected data as simple as possible).

Using the 'package' argument with a super simple package.json when calling the new e2e test, this additional test can be kept quite simple.

The package.json for this test should be placed into a subdirectory of the test/fixture directory (e.g. 'test/fixture/e2e.InstalledFrom') with content
``` json
{
  "dependencies": {
    "@angular/core": "^13.0.0",
  }
}
``` 
Unluckily we would also need a minimal package-lock.json with something like
``` json
{
  "dependencies": {
    "@angular/core":  {
      "version": "13.0.0",
      "resolved": "https://registry.npmjs.org/@angular/core/-/core-13.0.0.tgz",
    }
  }
}
``` 

In the 'endToEnd.test.js' we just nee a new test case:
``` js
const packagePathForInstalledFrom = path
	.resolve(__dirname, 'fixture/e2e.InstalledFrom', 'package.json')
	.replace(/(\s+)/g, '\\$1')
:
describe('end to end test for optional fields', function () {
	this.timeout(50000)

	it('produce a json report with field "installedFrom"', async () => {
		const { stdout, stderr } = await execAsPromise(`node ${scriptPath} --package=${packagePathForInstalledFrom }`)
		const result = JSON.parse(stdout)
		const expectedLengthOfResult = 1

		assert.strictEqual(result.length, expectedLengthOfResult, `expected the list to contain ${expectedLengthOfResult} elements`)
                // and what other asserts are needed
	});
}
```

The disadvantage would be that all your hard work on the tests would have been in vain :-(
And yes, I should think about simplifying the tests.

coyoteecd replied:
I force-pushed a set of changes, I modified heavily the way e2e test data is generated, breaking away the dependency between e2e tests and the actual deps of license-report (so that the test won't break when you update deps of the package itself).

Can be done better, but at this point I feel like I'm refactoring the original package's tests instead of adding a feature. Feel free to improve further upon merge.

BePo65 commented on test/getInstalledVersions.test.js:
```diff
@@ -1,28 +1,36 @@
   });
+
+  it('gets the installed from for installed packages', () => {
+    const packageLockContent = { dependencies: { mocha: { version: '6.2.3', resolved: 'https://registry.npmjs.org/mocha/-/mocha-6.2.3.tgz' } } }
```
As the new field is part of the data returned by getInstalledPackagesData imo we should validate the complete object returned by getInstalledPackagesData in one test.

So adding this line (and the corresponding assert) to the first test, we could drop this separate test.
And the 2 other test in this suite handle special cases of semver, where 'installedFrom' does not matter.

coyoteecd replied:
matter of taste I guess, I like tests a bit more focused.

I changed it anyway, forgot to rename the file too

BePo65 replied:
Just as an aside:
ss I am always interested in how others handle things: would you add a separate test for every property of an object returned by  a function?
I can imagine that this has the advantage of making clear what a test is for. But doesn't this generate an 'overwhelming' number of tests (and if so, is this a problem)?

coyoteecd replied:
No, I make a judgment call, making separate tests for groups of properties that logically belong together.

BePo65 reviewed (changes requested):
Sorry that I have one more point: let's make the new field **optional** to avoid a breaking change.

On the other hand, making `installedFrom` an optional field makes it easy to simplify the e2e tests; imo we just need 1 test to show that we can add `installedFrom` to the json output.

I put the details into my comments.

Thanks for your work (and you are really fast in responding to proposed changes :-)

coyoteecd committed:
feat: add support for installedFrom field

This displays the "resolved" URL for each package, as extracted from package-lock.json
Also rework end-to-end tests to use different package.json files for default/all fields configs
